apps/test-app/frontend/src/components/tree-widget/stateless-v2/TreeActions.ts

```diff
@@ -0,0 +1,292 @@
+      const modelNode = model.idToNode[nodeId];
+      if (modelNode && isTreeModelHierarchyNode(modelNode) && modelNode.children && model.parentChildMap.get(modelNode.id) === undefined) {
+        modelNode.isLoading = true;
+        loadChildren = true;
+      }
+    });
+
+    if (loadChildren) {
+      this.loadNodes(nodeId);
+    }
+  }
+
+  public setHierarchyLimit(nodeId: string | undefined, limit?: number | "unbounded") {
+    this.updateTreeModel((model) => {
+      removeSubTree(model, nodeId);
+      if (nodeId === undefined) {
+        model.rootNode.hierarchyLimit = limit;
+        return;
+      }
+
+      const modelNode = model.idToNode[nodeId];
+      if (modelNode && isTreeModelHierarchyNode(modelNode)) {
+        modelNode.hierarchyLimit = limit;
+        modelNode.isLoading = true;
+      }
+    });
+
+    this.loadNodes(nodeId);
+  }
+
+  public reloadTree(parentId: string | undefined, options?: { discardState?: boolean }) {
+    const oldModel = this._currentModel;
+    const expandedNodes = !!options?.discardState ? [] : collectNodes(parentId, oldModel, (node) => node.isExpanded === true);
+    const collapsedNodes = !!options?.discardState ? [] : collectNodes(parentId, oldModel, (node) => node.isExpanded === false);
+    const buildNode = !!options?.discardState ? (node: TreeModelHierarchyNode) => node : (node: TreeModelHierarchyNode) => addAttributes(node, oldModel);
+
+    const rootNode = parentId !== undefined ? this.getNode(parentId) : oldModel.rootNode;
+    if (!rootNode) {
+      return;
+    }
+
+    if (parentId === undefined) {
+      // cancel all ongoing requests
+      this._disposed.next();
+      this.updateTreeState((state) => ({ ...state, isLoading: true }));
+    }
+
+    this._loader
+      .reloadNodes(rootNode, { expandedNodes, collapsedNodes, buildNode })
+      .pipe(
+        takeUntil(this._disposed),
+        reduce<LoadedHierarchyPart, TreeModel>(
+          (treeModel, loadedPart) => {
+            addNodesToModel(treeModel, loadedPart);
+            const node = loadedPart.parent.id ? treeModel.idToNode[loadedPart.parent.id] : undefined;
+            // expand parent node
+            if (node && isTreeModelHierarchyNode(node)) {
+              node.isExpanded = true;
+            }
+            return treeModel;
+          },
+          {
+            idToNode: {},
+            parentChildMap: new Map(),
+            rootNode: !!options?.discardState ? { id: undefined, nodeData: undefined } : { ...oldModel.rootNode },
+          },
+        ),
+      )
+      .subscribe({
+        next: (newModel) => {
+          this.updateTreeModel((model) => {
+            addHierarchyPart(model, parentId, newModel);
+            const parentNode = parentId !== undefined ? model.idToNode[parentId] : undefined;
+            if (parentNode && isTreeModelHierarchyNode(parentNode)) {
+              parentNode.isLoading = false;
+            }
+          });
+        },
+      });
```

Comment:
I would very much like at least some of this to be shared across `reloadTree` and `loadNodes`

Reply:
Created operator `collectHierarchyPartsUntil`.